src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc

```diff
@@ -86,8 +88,33 @@ typedef struct grpc_ares_hostbyname_request {
   uint16_t port;
   /** is it a grpclb address */
   bool is_balancer;
+  /** for logging and errors: the query type ("A" or "AAAA") */
+  grpc_core::UniquePtr<char> qtype;
 } grpc_ares_hostbyname_request;
 
+static void grpc_ares_request_ref_locked(grpc_ares_request* r);
+static void grpc_ares_request_unref_locked(grpc_ares_request* r);
+
+class GrpcAresQuery {
```

Comment:
I don't understand the purpose of this class.  It seems to be a simple wrapper of `grpc_ares_request`, but it doesn't seem to provide any functionality that isn't already present in the latter.  Why is this wrapper being added?

Reply:
My idea here was that a `GrpcAresQuery` tracks fields that are particular to a single c-ares query (or "search") - e.g. a single A, AAAA, SRV, or TXT resolution performed by the c-ares APIs. A `grpc_ares_request`, on the other hand, tracks fields that are necessary to drive a complete DNS resolution process - which consists of multiple "child" queries.

So, the main thing is that a `grpc_ares_request` and a `GrpcAresQuery` differ in scope and lifetime - the former is per-grpc-resolution and the latter is per-c-ares-query.

Alternatively we could just put the SRV target name and TXT target name as additional fields into the `grpc_ares_request`, but keeping them separate seemed cleaner (I think this will become more important after making `grpc_ares_hostbyname_request` a sub-type of `GrpcAresQuery`).

Reply:
Okay, I guess this is fine for now.  I would like to find a cleaner way to do this that does not require allocating so many different data structures, but I guess we can optimize that as part of the larger C++ conversion.  For now, please add a TODO about this.